File: examples/comparison/dungeon/dungeon.go
// Same logic as App/NyraApp (Dungeon Steps) for fair runtime comparison.
package main

import "fmt"

const (
	grid        = 4
	startHP     = 100
	maxTurns    = 24
	trapDamage  = 25
	goldValue   = 15
	goalBonus   = 50
	appTitle    = "Dungeon Steps"
	statusPlaying = 0
	statusWon     = 1
	statusLost    = 2
)

type tile int

const (
	tileFloor tile = iota
	tileTrap
	tileGold
	tileGoal
)

type dir int

const (
	dirNorth dir = iota
	dirEast
	dirSouth
	dirWest
)

type gameState struct {
	x, y, hp, score, turn, status int32
}

func tileAt(x, y int32) tile {
	if x == 3 && y == 3 {
		return tileGoal
	}
	if x == 1 && y == 1 {
		return tileTrap
	}
	if x == 2 && y == 1 {
		return tileGold
	}
	if x == 1 && y == 3 {
		return tileGold
	}
	return tileFloor
}

func clampAxis(v int32) int32 {
	if v < 0 {
		return 0
	}
	if v >= grid {
		return grid - 1
	}
	return v
}

func turnPhase(turn int32) int32 {
	return turn % 4
}

func dirForTurn(turn int32) dir {
	switch turnPhase(turn) {
	case 0, 1:
		return dirEast
	default:
		return dirSouth
	}
}

func newGame() gameState {
	return gameState{x: 0, y: 0, hp: startHP, score: 0, turn: 0, status: statusPlaying}
}

func isPlaying(game gameState) int32 {
	if game.status == statusPlaying {
		return 1
	}
	return 0
}

func copyGame(x, y, hp, score, turn, status int32) gameState {
	return gameState{x: x, y: y, hp: hp, score: score, turn: turn, status: status}
}

func trapHP(hp int32) int32 {
	h := hp - trapDamage
	if h <= 0 {
		return 0
	}
	return h
}

func trapStatus(hp, playing int32) int32 {
	h := hp - trapDamage
	if h <= 0 {
		return statusLost
	}
	return playing
}

func applyTileAt(sx, sy, shp, sscore, sturn, sstatus int32, t tile) gameState {
	switch t {
	case tileFloor:
		return copyGame(sx, sy, shp, sscore, sturn, sstatus)
	case tileGold:
		return copyGame(sx, sy, shp, sscore+goldValue, sturn, sstatus)
	case tileGoal:
		return copyGame(sx, sy, shp, sscore+goalBonus, sturn, statusWon)
	case tileTrap:
		return copyGame(sx, sy, trapHP(shp), sscore, sturn, trapStatus(shp, sstatus))
	default:
		return copyGame(sx, sy, shp, sscore, sturn, sstatus)
	}
}

func advanceParts(sx, sy, shp, sscore, sturn, sstatus, dx, dy int32) gameState {
	nx := clampAxis(sx + dx)
	ny := clampAxis(sy + dy)
	return applyTileAt(nx, ny, shp, sscore, sturn+1, sstatus, tileAt(nx, ny))
}

func movePlayer(game gameState, d dir) gameState {
	sx, sy := game.x, game.y
	shp, sscore := game.hp, game.score
	sturn, sstatus := game.turn, game.status
	switch d {
	case dirEast:
		return advanceParts(sx, sy, shp, sscore, sturn, sstatus, 1, 0)
	case dirWest:
		return advanceParts(sx, sy, shp, sscore, sturn, sstatus, -1, 0)
	case dirNorth:
		return advanceParts(sx, sy, shp, sscore, sturn, sstatus, 0, -1)
	case dirSouth:
		return advanceParts(sx, sy, shp, sscore, sturn, sstatus, 0, 1)
	default:
		return game
	}
}

func stepTurn(game gameState) gameState {
	sx, sy := game.x, game.y
	shp, sscore := game.hp, game.score
	sturn, sstatus := game.turn, game.status
	if sstatus == statusPlaying {
		switch dirForTurn(sturn) {
		case dirEast:
			return advanceParts(sx, sy, shp, sscore, sturn, sstatus, 1, 0)
		case dirWest:
			return advanceParts(sx, sy, shp, sscore, sturn, sstatus, -1, 0)
		case dirNorth:
			return advanceParts(sx, sy, shp, sscore, sturn, sstatus, 0, -1)
		case dirSouth:
			return advanceParts(sx, sy, shp, sscore, sturn, sstatus, 0, 1)
		}
	}
	return copyGame(sx, sy, shp, sscore, sturn, sstatus)
}

func runDungeon() int32 {
	game := newGame()
	for isPlaying(game) != 0 && game.turn < maxTurns {
		game = stepTurn(game)
	}
	if isPlaying(game) == 0 {
		return game.score
	}
	lost := copyGame(game.x, game.y, game.hp, game.score, game.turn, statusLost)
	return lost.score
}

func outcomeCode(game gameState) int32 {
	return game.status
}

func main() {
	fmt.Println(appTitle)
	fmt.Println(runDungeon())
	g0 := newGame()
	g1 := movePlayer(g0, dirEast)
	fmt.Println(outcomeCode(g1))
	var banner int32
	for i := int32(0); i < 3; i++ {
		banner = banner + i
	}
	fmt.Println(banner)
}

File: examples/comparison/dungeon/dungeon.c
/* Same logic as App/NyraApp (Dungeon Steps) for fair runtime comparison. */
#include <stdio.h>
#include <stdint.h>

enum {
    GRID = 4,
    START_HP = 100,
    MAX_TURNS = 24,
    TRAP_DAMAGE = 25,
    GOLD_VALUE = 15,
    GOAL_BONUS = 50,
    STATUS_PLAYING = 0,
    STATUS_WON = 1,
    STATUS_LOST = 2,
    TILE_FLOOR = 0,
    TILE_TRAP = 1,
    TILE_GOLD = 2,
    TILE_GOAL = 3,
    DIR_NORTH = 0,
    DIR_EAST = 1,
    DIR_SOUTH = 2,
    DIR_WEST = 3
};

typedef struct {
    int32_t x, y, hp, score, turn, status;
} GameState;

static int32_t tile_at(int32_t x, int32_t y) {
    if (x == 3 && y == 3) return TILE_GOAL;
    if (x == 1 && y == 1) return TILE_TRAP;
    if (x == 2 && y == 1) return TILE_GOLD;
    if (x == 1 && y == 3) return TILE_GOLD;
    return TILE_FLOOR;
}

static int32_t clamp_axis(int32_t v) {
    if (v < 0) return 0;
    if (v >= GRID) return GRID - 1;
    return v;
}

static int32_t turn_phase(int32_t turn) { return turn % 4; }

static int32_t dir_for_turn(int32_t turn) {
    int32_t p = turn_phase(turn);
    if (p == 0 || p == 1) return DIR_EAST;
    return DIR_SOUTH;
}

static GameState new_game(void) {
    GameState g = {0, 0, START_HP, 0, 0, STATUS_PLAYING};
    return g;
}

static int32_t is_playing(GameState game) {
    return game.status == STATUS_PLAYING ? 1 : 0;
}

static GameState copy_game(int32_t x, int32_t y, int32_t hp, int32_t score, int32_t turn, int32_t status) {
    GameState g = {x, y, hp, score, turn, status};
    return g;
}

static int32_t trap_hp(int32_t hp) {
    int32_t h = hp - TRAP_DAMAGE;
    return h <= 0 ? 0 : h;
}

static int32_t trap_status(int32_t hp, int32_t playing) {
    int32_t h = hp - TRAP_DAMAGE;
    return h <= 0 ? STATUS_LOST : playing;
}

static GameState apply_tile_at(int32_t sx, int32_t sy, int32_t shp, int32_t sscore, int32_t sturn, int32_t sstatus, int32_t t) {
    switch (t) {
        case TILE_FLOOR: return copy_game(sx, sy, shp, sscore, sturn, sstatus);
        case TILE_GOLD: return copy_game(sx, sy, shp, sscore + GOLD_VALUE, sturn, sstatus);
        case TILE_GOAL: return copy_game(sx, sy, shp, sscore + GOAL_BONUS, sturn, STATUS_WON);
        case TILE_TRAP: return copy_game(sx, sy, trap_hp(shp), sscore, sturn, trap_status(shp, sstatus));
        default: return copy_game(sx, sy, shp, sscore, sturn, sstatus);
    }
}

static GameState advance_parts(int32_t sx, int32_t sy, int32_t shp, int32_t sscore, int32_t sturn, int32_t sstatus, int32_t dx, int32_t dy) {
    int32_t nx = clamp_axis(sx + dx);
    int32_t ny = clamp_axis(sy + dy);
    return apply_tile_at(nx, ny, shp, sscore, sturn + 1, sstatus, tile_at(nx, ny));
}

static GameState move_player(GameState game, int32_t d) {
    int32_t sx = game.x, sy = game.y;
    int32_t shp = game.hp, sscore = game.score;
    int32_t sturn = game.turn, sstatus = game.status;
    switch (d) {
        case DIR_EAST: return advance_parts(sx, sy, shp, sscore, sturn, sstatus, 1, 0);
        case DIR_WEST: return advance_parts(sx, sy, shp, sscore, sturn, sstatus, -1, 0);
        case DIR_NORTH: return advance_parts(sx, sy, shp, sscore, sturn, sstatus, 0, -1);
        case DIR_SOUTH: return advance_parts(sx, sy, shp, sscore, sturn, sstatus, 0, 1);
        default: return game;
    }
}

static GameState step_turn(GameState game) {
    int32_t sx = game.x, sy = game.y;
    int32_t shp = game.hp, sscore = game.score;
    int32_t sturn = game.turn, sstatus = game.status;
    if (sstatus == STATUS_PLAYING) {
        switch (dir_for_turn(sturn)) {
            case DIR_EAST: return advance_parts(sx, sy, shp, sscore, sturn, sstatus, 1, 0);
            case DIR_WEST: return advance_parts(sx, sy, shp, sscore, sturn, sstatus, -1, 0);
            case DIR_NORTH: return advance_parts(sx, sy, shp, sscore, sturn, sstatus, 0, -1);
            case DIR_SOUTH: return advance_parts(sx, sy, shp, sscore, sturn, sstatus, 0, 1);
        }
    }
    return copy_game(sx, sy, shp, sscore, sturn, sstatus);
}

static int32_t run_dungeon(void) {
    GameState game = new_game();
    while (is_playing(game) && game.turn < MAX_TURNS) {
        game = step_turn(game);
    }
    if (!is_playing(game)) return game.score;
    GameState lost = copy_game(game.x, game.y, game.hp, game.score, game.turn, STATUS_LOST);
    return lost.score;
}

static int32_t outcome_code(GameState game) { return game.status; }

int main(void) {
    printf("Dungeon Steps\n");
    printf("%d\n", run_dungeon());
    GameState g0 = new_game();
    GameState g1 = move_player(g0, DIR_EAST);
    printf("%d\n", outcome_code(g1));
    int32_t banner = 0;
    for (int32_t i = 0; i < 3; i++) banner += i;
    printf("%d\n", banner);
    return 0;
}
